File: tree/rooted-tree.go
// AOJ ALDS1_7_A: http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ALDS1_7_A

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var sc = bufio.NewScanner(os.Stdin)
var NULL = -1

type node struct {
	parent, left, right, depth int
}

func nextInt() int {
	sc.Scan()
	v, _ := strconv.Atoi(sc.Text())
	return v
}

// 再帰的に深さを求める
func recursion(nodes []node, n, p int) {
	nodes[n].depth = p
	if nodes[n].right != NULL {
		recursion(nodes, nodes[n].right, p)
	}

	if nodes[n].left != NULL {
		recursion(nodes, nodes[n].left, p+1)
	}
}

func getNodeType(n node) string {
	if n.parent == NULL {
		return "root"
	}

	if n.left == NULL {
		return "leaf"
	}

	return "internal node"
}

func main() {
	sc.Split(bufio.ScanWords)
	n := nextInt()

	nodes := make([]node, n)
	for i := 0; i < n; i++ {
		nodes[i] = node{NULL, NULL, NULL, NULL}
	}

	// structing tree
	var l int
	for i := 0; i < n; i++ {
		id := nextInt()
		k := nextInt()
		for j := 0; j < k; j++ {
			c := nextInt()
			if j == 0 {
				nodes[id].left = c
			} else {
				nodes[l].right = c
			}
			l = c
			nodes[c].parent = id
		}
	}

	// find root node
	var root int
	for i := 0; i < n; i++ {
		if nodes[i].parent == NULL {
			root = i
			break
		}
	}

	recursion(nodes, root, 0)
	for i, v := range nodes {
		var childlen []string
		for i := nodes[i].left; i != NULL; i = nodes[i].right {
			childlen = append(childlen, strconv.Itoa(i))
		}
		childlenStr := strings.Join(childlen, ", ")
		nodeType := getNodeType(v)
		fmt.Printf("node %d: parent = %d, depth = %d, %s, %s\n", i, v.parent, v.depth, nodeType, "["+childlenStr+"]")
	}
}

File: tree/rooted_tree.py
# coding: utf-8

# AOJ ALDS1_7_A: http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ALDS1_7_A

NULL = -1

class Node():
    def __init__(self):
        self.parent =  NULL # parent node
        self.left = NULL    # first child node
        self.right = NULL   # brother node
        self.depth = NULL   # depth of node
        self.type = ""      # type of node(root, internal node, leaf)

def set_depth(T: list, r: int, d: int):
    T[r].depth = d
    if T[r].left != NULL:
        set_depth(T, T[r].left, d + 1)

    if T[r].right != NULL:
        set_depth(T, T[r].right, d)

n = int(input())

T = []
for i in range(n):
    T.append(Node())

cur = 0
for i in range(n):
    nodeinfo = [int(x) for x in input().split()]
    for j in range(nodeinfo[1]):
        if j == 0:
            T[nodeinfo[0]].left = nodeinfo[2+j]
        else:
            T[cur].right = nodeinfo[2+j]
        cur = nodeinfo[2+j]
        T[cur].parent = nodeinfo[0]

# find root node
root = 0
for i, v in enumerate(T):
    if v.parent == NULL:
        root = i
        break

# set depth to all nodes
# set_depth(T, root, 0)

# set type
for v in T:
    if v.parent == NULL:
        v.type = "root"
    elif v.left == NULL:
        v.type = "leaf"
    else:
        v.type = "internal node"

for i, v in enumerate(T):
    childlen = []
    if T[i].left != NULL:
        child = T[i].left
        childlen.append(str(child))

        while T[child].right != NULL:
            child = T[child].right
            childlen.append(str(child))

    depth = 0
    cur = i
    while T[cur].parent != NULL:
        depth += 1
        cur = T[cur].parent
            
    print("node {}: parent = {}, depth = {}, {}, {}".format(i, v.parent, depth, v.type, "[" + ", ".join(childlen) + "]"))
